Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts

```diff
@@ -183,59 +217,86 @@ async function run() {
-            for (const fileToCopy of filesToCopy) {
-                try {
-                    tl.debug('fileToCopy = ' + fileToCopy);
-
-                    let relativePath;
-                    if (flattenFolders) {
-                        relativePath = path.basename(fileToCopy);
-                    } else {
-                        relativePath = fileToCopy.substring(sourceFolder.length)
-                            .replace(/^\\/g, "")
-                            .replace(/^\//g, "");
-                    }
-                    tl.debug('relativePath = ' + relativePath);
-                    let targetPath = path.posix.join(targetFolder, relativePath);
+        if (contents.length === 1 && contents[0] === "**") {
+            tl.debug("Upload a directory to a remote machine");
+
+            try {
+                const completedDirectory = await sshHelper.uploadFolder(sourceFolder, targetFolder);
+                console.log(tl.loc('CopyDirectoryCompleted', completedDirectory));
+            } catch (err) {
+                throw tl.loc("CopyDirectoryFailed", sourceFolder, err);
+            }
+        } else {
+            // identify the files to copy
+            const filesToCopy = getFilesToCopy(sourceFolder, contents);
 
-                    if (!path.isAbsolute(targetPath) && !utils.pathIsUNC(targetPath)) {
-                        targetPath = `./${targetPath}`;
+            // copy files to remote machine
+            if (filesToCopy) {
+                const preparedFiles = prepareFiles(filesToCopy, sourceFolder, targetFolder, flattenFolders);
+
+                tl.debug('Number of files to copy = ' + preparedFiles.length);
+                tl.debug('filesToCopy = ' + preparedFiles);
+
+                console.log(tl.loc('CopyingFiles', preparedFiles.length));
+
+                // Create remote folders structure
+                const folderStructure = getUniqueFolders(preparedFiles.map(x => x[1]).sort());
+
+                for (const foldersPath of folderStructure) {
+                    try {
+                        await sshHelper.createRemoteDirectory(foldersPath);
+                        console.log(tl.loc("FolderCreated", foldersPath));
+                    } catch (error) {
+                        throw tl.loc('TargetNotCreated', foldersPath, error);
                     }
+                }
+
+                console.log(tl.loc("FoldersCreated", folderStructure.length));
+
+                // Upload files to remote machine
+                const chunkSize = 10;
+                const chunks = Math.ceil(preparedFiles.length / chunkSize);
+                let successedFiles = 0;
+
+                for (let i = 0; i < chunks; i++) {
+                    // Splice files to copy into chunks
+                    const chunk = preparedFiles.slice(i * chunkSize, i * chunkSize + chunkSize);
+
+                    const results = await Promise.allSettled(chunk.map(async (pathTuple) => {
```

Comment:
looks like we queue up ten files to upload at same time and wait for all 10 to finish uploading?  Shouldn't this be a continous queue?

Reply:
The mechanism has been redesigned into a continuous queue